### PR TITLE
Check for dependencies before starting interactive prompt

### PR DESCRIPTION
**Pull request by chetan-rns**
**What type of PR is this?**

> /kind feature

**What does this PR do / why we need it**:

* This PR will add a check to confirm if the required dependencies are installed and provide the relevant instructions if it fails to satisfy the requirements. 
* Ask for sealed secret details only in the absence of default resources
* Fail if ArgoCD and Pipelines operator is not installed

**Have you updated the necessary documentation?**

See https://github.com/rhd-gitops-example/docs

[ ] Updated

**Which issue(s) this PR fixes**:

Fixes https://issues.redhat.com/browse/GITOPS-246

**How to test changes / Special notes to the reviewer**:

### CODE REVIEW

File: pkg/odo/cli/pipelines/utility/utility.go [hunk 1]
```diff
@@ -35,3 +38,47 @@ func MaybeCompletePrefix(s string) string {
...
+	KubeClient kubernetes.Interface
+}
+
+// NewClient returns a new K8s client
+func NewClient(client kubernetes.Interface) *Client {
+	return &Client{KubeClient: client}
+}
+
+// CheckIfSealedSecretsExists checks if sealed secrets is installed
+func (c *Client) CheckIfSealedSecretsExists(ns, name string) error {
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/utility/utility.go]: making this a `types.NamespacedName` makes it less likely to make mistakes.

File: pkg/odo/cli/pipelines/wizard.go [hunk 2]
```diff
@@ -90,6 +114,46 @@ func (io *WizardParameters) Complete(name string, cmd *cobra.Command, args []str
 	return nil
 }
 
+func checkBootstrapDependencies(io *WizardParameters, kubeClient kubernetes.Interface) error {
+
+	client := utility.NewClient(kubeClient)
+	log.Progressf("\nChecking dependencies\n")
+
+	sealedSpinner := log.Spinner("Checking if Sealed Secrets is installed at kube-system namespace")
+	err := client.CheckIfSealedSecretsExists(sealedSecretsNS+"s", sealedSecretsName)
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/wizard.go]: This (and the other case) should be able to determine if the k8s objects are "not found" vs "unable to contact the cluster" and report better messages.

File: pkg/odo/cli/pipelines/wizard.go [hunk 3]
```diff
@@ -90,6 +114,46 @@ func (io *WizardParameters) Complete(name string, cmd *cobra.Command, args []str
...
+		sealedSpinner.End(false)
+	} else {
+		io.SealedSecretsService.Name = sealedSecretsName
+		io.SealedSecretsService.Namespace = sealedSecretsNS
+		sealedSpinner.End(true)
+	}
+
+	argoSpinner := log.Spinner("Checking if ArgoCD Operator is installed at argocd namespace")
+	err = client.CheckIfArgoCDExists(argoCDNS)
+	if err != nil {
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/wizard.go:145]: Do we actually require ArgoCD now?

**Review by bigkevmcd — CHANGES_REQUESTED** [review]
I think we should do a better job of deciding whether or not the requested resource is "not found" in which case, installing it will help, vs "failed to communicate with a running cluster", in which case, we should report a better message.

File: pkg/odo/cli/pipelines/utility/utility.go (hunk 1 above)
Comment (chetan-rns) [reply pkg/odo/cli/pipelines/utility/utility.go]: Updated to use `types.NamespacedName`

File: pkg/odo/cli/pipelines/wizard.go (hunk 2 above)
Comment (chetan-rns) [reply pkg/odo/cli/pipelines/wizard.go]: Updated to throw different errors for "not found" vs "other cluster errors"

File: pkg/odo/cli/pipelines/wizard.go (hunk 3 above)
Comment (chetan-rns) [reply pkg/odo/cli/pipelines/wizard.go:145]: We add the excluded resources (taskruns and pipelineruns) to the ArgoCD operator resource. Applying the generated resources without installing the ArgoCD operator will throw an error. Hence we're checking if it's installed

**Comment by chetan-rns** [comment]
I've tried to mock the functionality of the Spinner using `io.Writer`. Unit tests will use a buffer to validate the content

File: pkg/odo/cli/pipelines/wizard.go [hunk 4]
```diff
@@ -90,6 +123,59 @@ func (io *WizardParameters) Complete(name string, cmd *cobra.Command, args []str
 	return nil
 }
 
+func checkBootstrapDependencies(io *WizardParameters, kubeClient kubernetes.Interface, spinner status) error {
+
+	var errs []error
+	client := utility.NewClient(kubeClient)
+	log.Progressf("\nChecking dependencies\n")
+
+	spinner.Start("Checking if Sealed Secrets is installed at kube-system namespace", false)
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/wizard.go]: How about "...is installed with the default configuration"?

File: pkg/odo/cli/pipelines/wizard.go [hunk 5]
```diff
@@ -90,6 +123,59 @@ func (io *WizardParameters) Complete(name string, cmd *cobra.Command, args []str
...
+		errs = append(errs, err)
+	}
+
+	if len(errs) > 0 {
+		return fmt.Errorf("Failed to satisfy the required dependencies")
+	}
+	return nil
+}
+
+func checkSpinner(spinner status, warningMsg string, err error) {
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/wizard.go]: `checkSpinner` doesn't really seem to describe what this does?

**Review by bigkevmcd — CHANGES_REQUESTED** [review]

File: pkg/odo/cli/pipelines/wizard.go (hunk 5 above)
Comment (chetan-rns) [reply pkg/odo/cli/pipelines/wizard.go]: I've changed the name to `setSpinnerStatus`

File: pkg/odo/cli/pipelines/wizard.go (hunk 4 above)
Comment (chetan-rns) [reply pkg/odo/cli/pipelines/wizard.go]: I've updated the spinner messages

**Review by bigkevmcd — APPROVED** [review]
Looks good!

**Pull request merged**